src/Akka.Streams.Kafka/Stages/Consumers/Abstract/SubSourceLogic.cs

```diff
@@ -450,191 +565,228 @@ public SubSourcePromiseControl(SourceShape<(
-                        subSourceCancelledCallback((
-                            topicPartition, 
-                            _buffer.TryPeek(out var next) ? new SeekToOffsetAndReEmit(next.Offset) : ReEmit.Instance));
-                        //CompleteStage();
-                    });
-                }
+                subSourceCancelledCallback((
+                    topicPartition, OnDownstreamFinishSubSourceCancellationStrategy));
+                base.InternalOnDownstreamFinish(ex);
+            });
+        }
+        
+        protected virtual ISubSourceCancellationStrategy OnDownstreamFinishSubSourceCancellationStrategy => _buffer.Count > 0 ? new SeekToOffsetAndReEmit(_buffer.Peek().Offset) : ReEmit.Instance;
 
-                public override void PreStart()
-                {
-                    base.PreStart();
-                    Log.Debug("{0} Starting SubSource for partition {1}", _actorNumber, _topicPartition);
-
-                    _subSourceActor = GetStageActor(MessageHandling());
-                    _subSourceActor.Watch(_consumerActor);
-                    
-                    _subSourceStartedCallback((_topicPartition, Control));
-                    // consumerActor.tell(RegisterSubStage(requestMessages.tps), subSourceActor.ref) // JVM
-                }
+        public override void PreStart()
+        {
+            base.PreStart();
+            Log.Info("{0} Starting SubSource for partition {1}", _actorNumber, _topicPartition);
 
-                private StageActorRef.Receive MessageHandling() => args =>
-                {
-                    var (actor, message) = args;
+            _subSourceActor = GetStageActor(MessageHandling());
+            _subSourceActor.Watch(_consumerActor);
+            
+            var controlAndActor = new ControlAndStageActor(Control, _subSourceActor.Ref);
+            var started = new SubSourceStageLogicControl(this._topicPartition, controlAndActor,
+                FilterRevokedPartitionAsyncCallback);
 
-                    switch (message)
-                    {
-                        case KafkaConsumerActorMetadata.Internal.Messages<K, V> messages:
-                            _requested = false;
-                            foreach (var consumerMessage in messages.MessagesList)
-                                _buffer.Enqueue(consumerMessage);
-
-                            Pump();
-                            break;
-                        case Status.Failure failure:
-                            FailStage(failure.Cause);
-                            break;
-                        case Terminated terminated when terminated.ActorRef.Equals(_consumerActor):
-                            FailStage(new ConsumerFailed());
-                            break;
-                    }
-                };
+            _subSourceStartedCallback(started);
+            _consumerActor.Tell(new KafkaConsumerActorMetadata.Internal.RegisterSubStage(_requestMessages.Topics),
```

Comment:
This is now implemented: we message the `KafkaConsumerActor` and let it know that we care about Topics X/Y/Z - that information will be stored inside of the `KafkaConsumerActor` to help ensure that messages don't get lost during rebalancing